README.md
# Icecast Stream Proxy

## About

This helps you to secure and lock down your icecast streams and stats.
You can configure the proxy to support only access from your radio station frontend website.

### Abilities

- relay a icecast stream and its stats meta data
- supports http basic auth, if your icecast server is configured for it
- can stream ogg and falls back to mp3 for microsoft browsers
- restrict stream and stats access for your radio station frontend website only

### Limitations

- ogg and mp3 icecast mount should serve the same content
- if http basic auth is used, same credentials must be used for all icecast streams

## Setup & Configuration

### Server

- `composer require dduers/icecast-stream-proxy`
- configuration is documented in `examples/example.php`

### Client

- you can embed the proxied stream to your website with this html5 code snipped
```
<audio preload="none">
    <source src="https://www.example.org/streamproxy">
</audio>
```

## Recommendations

- serve the icecast server, the icecast proxy and your radio station's website over `https://`
- (!) the http basic auth credentials are at risk when no secure https connection is uses

## Ressources

- https://icecast.org
- https://www.liquidsoap.info


CHANGELOG
# Changelog

## 1.0.1

- improved package documentation and example

## 1.0.0

- initial release


index.example.php
<?php
/**
 * Dduers\IcecastStreamProxy
 * by Daniel Duersteler, 2020, https://github.com/dduers
 * this is an example configuration
 */

/**
 * allow php script to run forever
 * this needs to be supported by your webserver and php installation
 * otherwise the stream will break after a certain few seconds
 */
set_time_limit (0);

/**
 * include the icecast stream proxy class
 */
require_once 'src/dduers/icecaststreamproxy.php';

/**
 * create class instance with your configuration
 */
$streamproxy = new \Dduers\IcecastStreamProxy(array(

    /**
     * the root url of your icecast server
     * change this, if your icecast server is not running on the same host as the 
     * icecast stream proxy does
     * note: no trailing '/'
     * default: 'http://localhost:8000'
     */
    'ic_url' => 'http://localhost:8000',

    /**
     * stats page of the ice cast server
     * the stats are available with /?status on the url where this script is hosted
     * note: leading '/'
     * default: '/status-json.xsl'
     */
    'ic_mount_stats' => '/status-json.xsl',

    /**
     * mime type of the stats
     * default: 'application/json'
     */
    'ic_mount_stats_mime' => 'application/json',
    
    /**
     * ogg encoded mount
     * default for all browsers as long no mp3 mount is specified
     * note: leading '/'
     * default: '/stream.ogg'
     */
    'ic_mount_ogg' => '/stream.ogg',
    
    /**
     * mp3 encoded mount
     * if speciefied, microsoft browsers fall back to it
     * note: leading '/'
     * default: NULL
     */
    'ic_mount_mp3' => '/stream.mp3',

    /**
     * http basic auth credentials
     * if one of those parameters is missing, http basic auth is disabled
     * note: the stats page and all streams must share the same credentials
     * note: your icecast server must be setup to supprt http basic auth
     * default: both NULL
     */
    'ic_basic_auth_username' => NULL,
    'ic_basic_auth_password' => NULL,
    
    /**
     * frontend website
     * if you give an url here, all stream and stats requests are
     * limited to this referer (uses http referer header)
     * note: no trailing '/'
     * default: NULL
     */
    'sp_lock_http_referer' => 'https://www.radiostream.ch',

    /**
     * name of the getvar to query stats
     * you can get the icecasts stats with a call to (proxy url)/?status
     * default: 'status'
     */
    'sp_stats_get_var' => 'status',

    /**
     * size in bytes of a single chunk to read
     * default: 1024 * 1024
     */
    'sp_stream_chunk_size' => 1024 * 1024
));

/**
 * uncomment this line, when you're ready
 */
//$streamproxy->run();


examples/example.php
<?php
/**
 * Dduers\IcecastStreamProxy
 * by Daniel Duersteler, http://www.xsite.ch
 * this is an example configuration and startup script
 */

/**
 * allow php script to run forever
 * this needs to be supported by your webserver and php installation
 * otherwise the stream will break after a certain few seconds or minutes
 */
set_time_limit(0);

/**
 * include the icecast stream proxy class
 */
require 'vendor/autoload.php';

/**
 * create class instance with your configuration
 */
$streamproxy = new \Dduers\IcecastStreamProxy(array(

    /**
     * the root url of your icecast server
     * change this, if your icecast server is not running on the same host as the 
     * icecast stream proxy does
     * note: no trailing '/'
     * default: 'http://localhost:8000'
     */
    'ic_url' => 'http://localhost:8000',

    /**
     * stats page of the ice cast server
     * the stats are available with /?status on the url where this script is hosted
     * note: leading '/'
     * default: '/status-json.xsl'
     */
    'ic_mount_stats' => '/status-json.xsl',

    /**
     * mime type of the stats
     * default: 'application/json'
     */
    'ic_mount_stats_mime' => 'application/json',
    
    /**
     * ogg encoded mount
     * default for all browsers as long no mp3 mount is specified
     * note: leading '/'
     * default: '/stream.ogg'
     */
    'ic_mount_ogg' => '/stream.ogg',
    
    /**
     * mp3 encoded mount
     * if speciefied, microsoft browsers fall back to it
     * note: leading '/'
     * default: NULL
     */
    'ic_mount_mp3' => '/stream.mp3',

    /**
     * http basic auth credentials
     * if one of those parameters is missing, http basic auth is disabled
     * note: the stats page and all streams must share the same credentials
     * note: your icecast server must be setup to supprt http basic auth
     * default: both NULL
     */
    'ic_basic_auth_username' => NULL,
    'ic_basic_auth_password' => NULL,
    
    /**
     * frontend website
     * if you give an url here, all stream and stats requests are
     * limited to this referer (uses http referer header)
     * note: no trailing '/'
     * default: NULL
     */
    'sp_lock_http_referer' => 'https://www.radiostream.ch',

    /**
     * name of the getvar to query stats
     * you can get the icecasts stats with a call to (proxy url)/?status
     * default: 'status'
     */
    'sp_stats_get_var' => 'status',

    /**
     * size in bytes of a single chunk to read
     * default: 1024 * 1024
     */
    'sp_stream_chunk_size' => 1024 * 1024
));

/**
 * start streaming or fallback on error
 */
if ($streamproxy->run() === false) {

    /**
     * your error handling comes here.
     * here are some examples:
     */

    /** 
     * option 1:
     * send a http error code
     * this is best and also suits for ajax requests to the proxy
     */
    //http_response_code(403);

    /** 
     * option 2:
     * output a readable error message
     */
    //echo "ACCESS_DENIED";

    /**
     * option 3:
     * display a custom error page
     */
    //include "403.html";

    die();
};


icecastStreamProxy.class.php
<?php
/**
 * icecastStreamProxy
 * by Daniel Duersteler, 2020, https://github.com/dduers
 */
class icecastStreamProxy
{
    private $config;
    private $defaults = array(
        'ic_url' => 'http://localhost:8000',
        'ic_mount_stats' => '/status-json.xsl',
        'ic_mount_stats_mime' => 'application/json',
        'ic_mount_ogg' => '/stream.ogg',
        'ic_mount_mp3' => NULL,
        'ic_basic_auth_username' => NULL,
        'ic_basic_auth_password' => NULL,
        'sp_lock_http_referer' => NULL,
        'sp_stats_get_var' => 'status',
        'sp_stream_chunk_size' => 1024 * 1024
    );
    private $stream_file = NULL;
    private $stream_mime = NULL;

    function __construct(array $options = array())
    {
        $this->config = array_merge($this->defaults, $options);
        if ($this->config['ic_url'] && ($this->config['ic_mount_ogg'] || $this->config['ic_mount_mp3'])) {
            if ($this->is_microsoft_browser()) {
                if ($this->config['ic_mount_mp3']) {
                    $this->stream_file = $this->config['ic_url'] . $this->config['ic_mount_mp3'];
                    $this->stream_mime = 'audio/mpeg';
                } elseif ($this->config['ic_mount_ogg']) { 
                    $this->stream_file = $this->config['ic_url'] . $this->config['ic_mount_ogg'];
                    $this->stream_mime = 'audio/ogg';
                }
            } else {
                if ($this->config['ic_mount_ogg']) {
                    $this->stream_file = $this->config['ic_url'] . $this->config['ic_mount_ogg'];
                    $this->stream_mime = 'audio/ogg';
                } elseif ($this->config['ic_mount_mp3']) {
                    $this->stream_file = $this->config['ic_url'] . $this->config['ic_mount_mp3'];
                    $this->stream_mime = 'audio/mpeg';
                }
            }
        }
    }
    
    function run()
    {
        if ($this->config['sp_lock_http_referer'] && !isset($_SERVER['HTTP_REFERER']))
            return false;
        if ($this->config['sp_lock_http_referer'] && (strpos($_SERVER['HTTP_REFERER'], $this->config['sp_lock_http_referer'].'/') === false))
            return false;
        if(isset($_GET[$this->config['sp_stats_get_var']]))
            return $this->ic_get_stats();
        return $this->ic_start_stream();
    }

    private function ic_start_stream()
    {
        if (!($this->stream_file && $this->stream_mime))
            return false;
        $context = NULL;
        if ($this->config['ic_basic_auth_username'] && $this->config['ic_basic_auth_password']) {
            $context = stream_context_create(array(
                'http' => array(
                    'header' => 'Authorization: Basic ' . base64_encode($this->config['ic_basic_auth_username'] . ':' . $this->config['ic_basic_auth_password'])
                )
            ));
            
        }
        $handle = fopen($this->stream_file, 'rb', false, $context);
        if ($handle === false)
            return false;
        if ($this->stream_mime)
            header('Content-Type: ' . $this->stream_mime);
        while (!feof($handle)) {
            echo fread($handle, $this->config['sp_stream_chunk_size']);
            ob_flush();
            flush();
        }
        return fclose($handle); 
    }

    private function ic_get_stats()
    {
        if (!($this->config['ic_url'] && $this->config['ic_mount_stats'] && $this->config['ic_mount_stats_mime']))
            return false;
        $context = NULL;
        if ($this->config['ic_basic_auth_username'] && $this->config['ic_basic_auth_password']) {
            $context = stream_context_create(array(
                'http' => array(
                    'header' => 'Authorization: Basic ' . base64_encode($this->config['ic_basic_auth_username'] . ':' . $this->config['ic_basic_auth_password'])
                )
            ));
        }
        $content = file_get_contents($this->config['ic_url'] . $this->config['ic_mount_stats'], false, $context);
        if ($content === false)
            return false;
        if ($this->config['sp_lock_http_referer'])
            header('Access-Control-Allow-Origin: ' . $this->config['sp_lock_http_referer']);
        else 
            header('Access-Control-Allow-Origin: *');
        if ($this->config['ic_mount_stats_mime'])
            header('Content-Type: ' . $this->config['ic_mount_stats_mime']);
        echo $content;
        return true;
    }
    
    private function is_microsoft_browser()
    {
        return 
            preg_match('/Edge/i', $_SERVER['HTTP_USER_AGENT']) 
            || preg_match('~MSIE|Internet Explorer~i', $_SERVER['HTTP_USER_AGENT']) || (strpos($_SERVER['HTTP_USER_AGENT'], 'Trident/7.0; rv:11.0') !== false);
    }
}


src/dduers/icecaststreamproxy.php
<?php
namespace Dduers;

/**
 * Dduers\IcecastStreamProxy
 * by Daniel Duersteler, http://www.xsite.ch
 */
class IcecastStreamProxy
{
    private $config;
    private $defaults = array(
        'ic_url' => 'http://localhost:8000',
        'ic_mount_stats' => '/status-json.xsl',
        'ic_mount_stats_mime' => 'application/json',
        'ic_mount_ogg' => '/stream.ogg',
        'ic_mount_mp3' => NULL,
        'ic_basic_auth_username' => NULL,
        'ic_basic_auth_password' => NULL,
        'sp_lock_http_referer' => NULL,
        'sp_stats_get_var' => 'status',
        'sp_stream_chunk_size' => 1024 * 1024
    );
    private $stream_file = NULL;
    private $stream_mime = NULL;

    function __construct(array $options = array())
    {
        $this->config = array_merge($this->defaults, $options);
        if ($this->config['ic_url'] && ($this->config['ic_mount_ogg'] || $this->config['ic_mount_mp3'])) {
            if ($this->is_microsoft_browser()) {
                if ($this->config['ic_mount_mp3']) {
                    $this->stream_file = $this->config['ic_url'] . $this->config['ic_mount_mp3'];
                    $this->stream_mime = 'audio/mpeg';
                } elseif ($this->config['ic_mount_ogg']) { 
                    $this->stream_file = $this->config['ic_url'] . $this->config['ic_mount_ogg'];
                    $this->stream_mime = 'audio/ogg';
                }
            } else {
                if ($this->config['ic_mount_ogg']) {
                    $this->stream_file = $this->config['ic_url'] . $this->config['ic_mount_ogg'];
                    $this->stream_mime = 'audio/ogg';
                } elseif ($this->config['ic_mount_mp3']) {
                    $this->stream_file = $this->config['ic_url'] . $this->config['ic_mount_mp3'];
                    $this->stream_mime = 'audio/mpeg';
                }
            }
        }
    }
    
    function run()
    {
        if ($this->config['sp_lock_http_referer'] && !isset($_SERVER['HTTP_REFERER']))
            return false;
        if ($this->config['sp_lock_http_referer'] && (strpos($_SERVER['HTTP_REFERER'], $this->config['sp_lock_http_referer'].'/') === false))
            return false;
        if(isset($_GET[$this->config['sp_stats_get_var']]))
            return $this->ic_get_stats();
        return $this->ic_start_stream();
    }

    private function ic_start_stream()
    {
        if (!($this->stream_file && $this->stream_mime))
            return false;
        $context = NULL;
        if ($this->config['ic_basic_auth_username'] && $this->config['ic_basic_auth_password']) {
            $context = stream_context_create(array(
                'http' => array(
                    'header' => 'Authorization: Basic ' . base64_encode($this->config['ic_basic_auth_username'] . ':' . $this->config['ic_basic_auth_password'])
                )
            ));
            
        }
        $handle = fopen($this->stream_file, 'rb', false, $context);
        if ($handle === false)
            return false;
        if ($this->stream_mime)
            header('Content-Type: ' . $this->stream_mime);
        while (!feof($handle)) {
            echo fread($handle, $this->config['sp_stream_chunk_size']);
            ob_flush();
            flush();
        }
        return fclose($handle); 
    }

    private function ic_get_stats()
    {
        if (!($this->config['ic_url'] && $this->config['ic_mount_stats'] && $this->config['ic_mount_stats_mime']))
            return false;
        $context = NULL;
        if ($this->config['ic_basic_auth_username'] && $this->config['ic_basic_auth_password']) {
            $context = stream_context_create(array(
                'http' => array(
                    'header' => 'Authorization: Basic ' . base64_encode($this->config['ic_basic_auth_username'] . ':' . $this->config['ic_basic_auth_password'])
                )
            ));
        }
        $content = file_get_contents($this->config['ic_url'] . $this->config['ic_mount_stats'], false, $context);
        if ($content === false)
            return false;
        if ($this->config['sp_lock_http_referer'])
            header('Access-Control-Allow-Origin: ' . $this->config['sp_lock_http_referer']);
        else 
            header('Access-Control-Allow-Origin: *');
        if ($this->config['ic_mount_stats_mime'])
            header('Content-Type: ' . $this->config['ic_mount_stats_mime']);
        echo $content;
        return true;
    }
    
    private function is_microsoft_browser()
    {
        return 
            preg_match('/Edge/i', $_SERVER['HTTP_USER_AGENT']) 
            || preg_match('~MSIE|Internet Explorer~i', $_SERVER['HTTP_USER_AGENT']) || (strpos($_SERVER['HTTP_USER_AGENT'], 'Trident/7.0; rv:11.0') !== false);
    }
}
